Issue #398: FileImageDescriptor and URLImageDescriptor should hold...

Dani-Hub opened the pull request:
... only a single ImageFileNameProvider instance

- Introduce a lazily generated ImageFileNameProvider in FileImageDescriptor and URLImageDescriptor.
- Make the pre-existing fields of FileImageDescriptor and URLImageDescriptor final as an additional guarantee that these are effectively immutable, justifying the usage of an only once generated ImageFileNameProvider.
- Extend existing tests for FileImageDescriptor and URLImageDescriptor that getAdapter returns a unique instance of ImageFileNameProvider.

Signed-off-by: Daniel Krügler <daniel.kruegler@gmail.com>

Dani-Hub committed:
Issue #398: FileImageDescriptor and URLImageDescriptor should hold...

... only a single ImageFileNameProvider instance

- Introduce a lazily generated ImageFileNameProvider in
FileImageDescriptor and URLImageDescriptor.
- Make the pre-existing fields of FileImageDescriptor and
URLImageDescriptor final as an additional guarantee that these are
effectively immutable, justifying the usage of an only once generated
ImageFileNameProvider.
- Extend existing tests for FileImageDescriptor and URLImageDescriptor
that getAdapter returns a unique instance of ImageFileNameProvider.

Signed-off-by: Daniel Krügler <daniel.kruegler@gmail.com>

Dani-Hub commented:
@laeubi Could you please take a look?

laeubi commented:
@Dani-Hub I have the strong-feeling that `FileImageDescriptor` can simply implements `ImageFileNameProvider`, same for `URLImageDescriptor`, why having separate inner classes at all?

Dani-Hub committed:
Replace inner ImageFileNameProvider classes by inheritance

Dani-Hub commented:
I replaced only the inner `ImageFileNameProvider` classes as part of this issue. We can replace the inner `ImageDataProvider`  of `URLImageDescriptor` as part of my additionally planned improvements as part of issue #399.

Dani-Hub committed:
Replace IAdaptable inheritance by Adapters.adapt

Dani-Hub commented:
I also decided to follow your earlier advice to use the Adapters.adapt approach of directly testing the `IAdaptable` inheritance. I agree that this makes the test more robust against a possible future refactoring (and a little bit shorter as well).

laeubi commented:
I think you can now even omit the `IAdaptable` interface at all as `Adapters.adapt` already do instanceof check!

Dani-Hub commented:
The reason for the `IAdaptable` pattern has been introduced to solve the larger picture issue where we want to adapt to URL. Once we have implemented that, we can remove the adaption to the ImageFileNameProvider, I agree. And in fact that's the reason why I switched to Adapters.adapt right *now*.

laeubi commented:
> has been introduced to solve the larger picture issue where we want to adapt to URL. Once we have implemented that, we can remove the adaption to the ImageFileNameProvider, I agree.

Can you then add the code to adapt to `File` or `URL` object here? `return adapter.cast(this);` is not required anymore if the class itself implement the interface. if you still want to keep this maybe 

```
if (adapter.isInstance(this)) {
	// Support testing ImageFileNameProvider characteristics, see #396
	return adapter.cast(this);
}
```

might be more generic

Dani-Hub commented:
> > has been introduced to solve the larger picture issue where we want to adapt to URL. Once we have implemented that, we can remove the adaption to the ImageFileNameProvider, I agree.
> 
> Can you then add the code to adapt to `File` or `URL` object here? 

Not as part of this issue, please. I have already created a separate issue for the URL adaption. I don't plan to adapt to File right now. If you plan to do that, I suggest a separate issue (I don't think that this is necessary)

> 
> ```
> if (adapter.isInstance(this)) {
> 	// Support testing ImageFileNameProvider characteristics, see #396
> 	return adapter.cast(this);
> }
> ```
 
If we strike the adaption to ImageFileNameProvider later, this if case will get removed anyway. I would therefore prefer to keep it until that point.

Dani-Hub commented:
@laeubi Tests completed.
